src/mem3/src/mem3_reshard.erl

```diff
@@ -0,0 +1,919 @@
+    code_change/3
+]).
+
+
+-include("mem3_reshard.hrl").
+
+
+-define(JOB_ID_VERSION, 1).
+-define(JOB_STATE_VERSION, 1).
+-define(DEFAULT_MAX_JOBS, 25).
+-define(DEFAULT_MAX_HISTORY, 20).
+-define(JOB_PREFIX, <<"reshard-job-">>).
+-define(STATE_PREFIX, <<"reshard-state-">>).
+
+
+%% Public API
+
+-spec start_link() -> {ok, pid()} | ignore | {error, term()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+-spec start() -> ok | {error, any()}.
+start() ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> gen_server:call(?MODULE, start, infinity)
+    end.
+
+
+-spec stop(binary()) -> ok | {error, any()}.
+stop(Reason) ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> gen_server:call(?MODULE, {stop, Reason}, infinity)
+    end.
+
+
+-spec start_split_job(#shard{} | binary()) -> {ok, binary()} | {error, term()}.
+start_split_job(#shard{} = Shard) ->
+    start_split_job(Shard, 2);
+
+start_split_job(ShardName) when is_binary(ShardName) ->
+    start_split_job(shard_from_name(ShardName), 2).
+
+
+-spec start_split_job(#shard{}, split()) -> {ok, binary()} | {error, any()}.
+start_split_job(#shard{} = Source, Split) ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> validate_and_start_job(Source, Split)
+    end.
+
+
+-spec stop_job(binary(), binary()) -> ok | {error, any()}.
+stop_job(JobId, Reason) when is_binary(JobId), is_binary(Reason) ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> gen_server:call(?MODULE, {stop_job, JobId, Reason}, infinity)
+    end.
+
+
+-spec resume_job(binary()) -> ok | {error, any()}.
+resume_job(JobId) when is_binary(JobId) ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> gen_server:call(?MODULE, {resume_job, JobId}, infinity)
+    end.
+
+
+-spec remove_job(binary()) -> ok | {error, any()}.
+remove_job(JobId) when is_binary(JobId) ->
+    case is_disabled() of
+        true -> {error, resharding_disabled};
+        false -> gen_server:call(?MODULE, {remove_job, JobId}, infinity)
+    end.
+
+
+-spec get_state() -> {[_ | _]}.
+get_state() ->
+    gen_server:call(?MODULE, get_state, infinity).
+
+
+-spec jobs() -> [[tuple()]].
+jobs() ->
+    ets:foldl(fun(Job, Acc) ->
+        Opts = [iso8601],
+        Props = mem3_reshard_store:job_to_ejson_props(Job, Opts),
+        [{Props} | Acc]
+    end, [], ?MODULE).
+
+
+-spec job(job_id()) -> {ok, {[_ | _]}} | {error, not_found}.
+job(JobId) ->
+    case job_by_id(JobId) of
+        #job{} = Job ->
+            Opts = [iso8601],
+            Props = mem3_reshard_store:job_to_ejson_props(Job, Opts),
+            {ok, {Props}};
+        not_found ->
+            {error, not_found}
+    end.
+
+
+% Return true resharding is disabled in the application level settings
+-spec is_disabled() -> boolean().
+is_disabled() ->
```

Comment:
I think the idea was that disabling it would be exceptional and eventually the `is_disabled` would be removed altogether. Since shards won't automatically split, this is more of an extra safety net that can be used initially in a production setting but then removed in a later release.